libfreerdp/core/nla.c

```diff
@@ -2049,91 +1968,142 @@ static BOOL nla_read_ts_credentials(rdpNla* nla, SecBuffer* data, size_t offset)
 
 	WINPR_ASSERT(nla);
 	WINPR_ASSERT(nla->rdpcontext);
 
 	settings = nla->rdpcontext->settings;
 	WINPR_ASSERT(settings);
 
+	enc = WinPrAsn1Encoder_New(WINPR_ASN1_DER);
+	if (!enc)
+		return FALSE;
+
+	/* TSCredentials */
+	if (!WinPrAsn1EncSeqContainer(enc))
+		goto out;
+
+	/* credType [0] INTEGER */
+	credType = settings->SmartcardLogon ? 2 : 1;
+	if (!WinPrAsn1EncContextualInteger(enc, 0, credType))
+		goto out;
+
+	/* credentials [1] OCTET STRING */
+	if (!WinPrAsn1EncContextualOctetStringContainer(enc, 1))
+		goto out;
+
 	if (settings->SmartcardLogon)
 	{
-		TSSmartCardCreds_t smartcardCreds = { 0 };
-		TSCspDataDetail_t cspData = { 0 };
-		char* Password = freerdp_settings_get_string_writable(settings, FreeRDP_Password);
+		struct
+		{
+			WinPrAsn1_tagId tag;
+			size_t setting_id;
+		} cspData_fields[] = { { 1, FreeRDP_CardName },
+			                   { 2, FreeRDP_ReaderName },
+			                   { 3, FreeRDP_ContainerName },
+			                   { 4, FreeRDP_CspName } };
+		WinPrAsn1_OctetString octet_string;
+		char* str;
+		BOOL ret;
+
+		/* TSSmartCardCreds */
+		if (!WinPrAsn1EncSeqContainer(enc))
+			goto out;
 
-		smartcardCreds.pin = Password ? Password : "";
+		/* pin [0] OCTET STRING */
+		str = freerdp_settings_get_string_writable(settings, FreeRDP_Password);
+		octet_string.len =
+		    ConvertToUnicode(CP_UTF8, 0, str, -1, (LPWSTR*)&octet_string.data, 0) * sizeof(WCHAR);
+		ret = WinPrAsn1EncContextualOctetString(enc, 0, &octet_string);
+		if (octet_string.data)
+			free(octet_string.data);
+		if (!ret)
+			goto out;
 
-		/*smartcardCreds.userHint = settings->UserHint;
-		smartcardCreds.domainHint = settings->DomainHint;*/
-		smartcardCreds.cspData = &cspData;
+		/* cspData [1] SEQUENCE */
+		if (!WinPrAsn1EncContextualSeqContainer(enc, 1))
+			goto out;
 
-		cspData.keySpec = freerdp_settings_get_uint32(settings, FreeRDP_KeySpec);
-		cspData.cspName = freerdp_settings_get_string_writable(settings, FreeRDP_CspName);
-		cspData.readerName = freerdp_settings_get_string_writable(settings, FreeRDP_ReaderName);
-		cspData.cardName = freerdp_settings_get_string_writable(settings, FreeRDP_CardName);
-		cspData.containerName =
-		    freerdp_settings_get_string_writable(settings, FreeRDP_ContainerName);
+		/* keySpec [0] INTEGER */
+		if (!WinPrAsn1EncContextualInteger(enc, 0,
+		                                   freerdp_settings_get_uint32(settings, FreeRDP_KeySpec)))
+			goto out;
 
-		length = ber_sizeof_nla_TSSmartCardCreds(&smartcardCreds);
-		credsContentStream = Stream_New(NULL, length);
-		if (!credsContentStream)
-			return FALSE;
+		for (int i = 0; i < ARRAYSIZE(cspData_fields); i++)
+		{
+			str = freerdp_settings_get_string_writable(settings, cspData_fields[i].setting_id);
+			octet_string.len =
+			    ConvertToUnicode(CP_UTF8, 0, str, -1, (LPWSTR*)&octet_string.data, 0) *
+			    sizeof(WCHAR);
+			ret = WinPrAsn1EncContextualOctetString(enc, cspData_fields[i].tag, &octet_string);
+			if (octet_string.len)
```

Comment:
Don't check for NULL before `free`

Reply:
Yes but `octet_string.data` wouldn't be NULL. If `ConvertToUnicode` didn't do the conversion for whatever reason, it would be the previously freed pointer from the last conversion.

Reply:
Actually `octet_string.data` should be initialized with `NULL` before the `ConvertToUnicode` call and then just `free`d.

You could just do a `WinPrAsn1_OctetString octet_string = { 0 };` to have the struct correctly initialized.